core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -31,19 +32,24 @@ import scala.util.{Failure, Success}
 /**
  * Internal API.
  *
- * Anonymous sub-class instance is created in [[CommittableSubSource]].
+ * Anonymous sub-class instance is created in [[CommittableSubSource]], [[PlainSubSource]] and [[TransactionalSubSource]].
  */
 @InternalApi
-private abstract class SubSourceLogic[K, V, Msg](
+private class SubSourceLogic[K, V, Msg](
     val shape: SourceShape[(TopicPartition, Source[Msg, NotUsed])],
     settings: ConsumerSettings[K, V],
     subscription: AutoSubscription,
     getOffsetsOnAssign: Option[Set[TopicPartition] => Future[Map[TopicPartition, Long]]] = None,
-    onRevoke: Set[TopicPartition] => Unit = _ => ()
+    onRevoke: Set[TopicPartition] => Unit = _ => (),
+    msgSourceLogicFactory: (SourceShape[Msg],
+                            TopicPartition,
+                            ActorRef,
+                            AsyncCallback[(TopicPartition, (Control, ActorRef))],
+                            AsyncCallback[(TopicPartition, Option[ConsumerRecord[K, V]])],
+                            Int) => MessageSubSourceLogic[K, V, Msg]
 ) extends TimerGraphStageLogic(shape)
```

Comment:
This is worth a named type `SourceLogicFactory`.

Reply:
I renamed `MessageSubSourceLogic` to `SubSourceStageLogic` and created a type called `SubSourceStageLogicFactory`.  I also implemented concrete types for Transactional, Committable, and Plain sub sources that subclass `SubSourceStageLogic`.